12-offer-encoding.md

```diff
@@ -0,0 +1,854 @@
+  
+Assume L1A2A > L3A:
+
+  L1A2A=H("LnBranch",L1A||L2A)          L3A
+                          \            /
+                           v          v
+                Root=H("LnBranch",L3A||L1A2A)
+
+Signature = SIG("lightninginvoicesignature", Root, nodekey)
+```
+
+# Offers
+
+Offers are a precursor to an invoice_request: readers will request an invoice
+(or multiple) based on the offer.  An offer can be much longer-lived than a
+particular invoice, so it has some different characteristics; in particular the amount can be in a non-lightning currency.  It's
+also designed for compactness to fit inside a QR code easily.
+
+Note that the non-signature TLV elements get mirrored into
+invoice_request and invoice messages, so they each have specific and
+distinct TLV ranges.
+
+The human-readable prefix for offers is `lno`.
+
+## TLV Fields for Offers
+
+1. `tlv_stream`: `offer`
+2. types:
+    1. type: 2 (`offer_chains`)
+    2. data:
+        * [`...*chain_hash`:`chains`]
+    1. type: 4 (`offer_metadata`)
+    2. data:
+        * [`...*byte`:`data`]
+    1. type: 6 (`offer_currency`)
+    2. data:
+        * [`...*utf8`:`iso4217`]
+    1. type: 8 (`offer_amount`)
+    2. data:
+        * [`tu64`:`amount`]
+    1. type: 10 (`offer_description`)
+    2. data:
+        * [`...*utf8`:`description`]
+    1. type: 12 (`offer_features`)
+    2. data:
+        * [`...*byte`:`features`]
+    1. type: 14 (`offer_absolute_expiry`)
+    2. data:
+        * [`tu64`:`seconds_from_epoch`]
+    1. type: 16 (`offer_paths`)
+    2. data:
+        * [`...*blinded_path`:`paths`]
+    1. type: 18 (`offer_issuer`)
+    2. data:
+        * [`...*utf8`:`issuer`]
+    1. type: 20 (`offer_quantity_max`)
+    2. data:
+        * [`tu64`:`max`]
+    1. type: 22 (`offer_node_id`)
+    2. data:
+        * [`point`:`node_id`]
+
+## Requirements For Offers
+
+A writer of an offer:
+  - MUST NOT set any tlv fields greater or equal to 80, or tlv field 0.
+  - if the chain for the invoice is not solely bitcoin:
+    - MUST specify `offer_chains` the offer is valid for.
+  - otherwise:
+    - MAY omit `offer_chains`, implying that bitcoin is only chain.
+  - if a specific minimum `offer_amount` is required for successful payment:
+    - MUST set `offer_amount` to the amount expected (per item).
+    - if the currency for `offer_amount` is that of all entries in `chains`:
+      - MUST specify `amount` in multiples of the minimum lightning-payable unit
+        (e.g. milli-satoshis for bitcoin).
+    - otherwise:
+      - MUST specify `offer_currency` `iso4217` as an ISO 4712 three-letter code.
+      - MUST specify `offer_amount` in the currency unit adjusted by the ISO 4712
+        exponent (e.g. USD cents).
+    - MUST set `offer_description` to a complete description of the purpose
+      of the payment.
+  - otherwise:
+    - MUST NOT set `offer_amount`
+    - MUST NOT set `offer_currency`
+    - MAY set `offer_description`
```

Comment:
Can this explicitly specify it should be the user's choice? Currently, Phoenix isn't giving users the option, so their wallet is incompatible with other BOLT12 services.

Reply:
The reason we're only exposing one offer right now in Phoenix is because we're deriving it entirely from the user's seed, which means it can be restored from seed on a different phone and keep receiving on that same Bolt 12 offer. If we allow users setting custom descriptions:

- it's a clunky UX that isn't well suited for a mobile phone
- it has to be stored in a DB on the phone, which may be lost when changing phone

It would make sense for `phoenixd` though, where we will likely add an API to generate arbitrary offers, and users can easily backup their sqlite DB.

It's great that we have you here to discuss this issue though: I'm struggling to understand why ocean requires this custom description. Can you explain in more details here why it's needed, so that we can see if there's a better way to make it work seamlessly with Bolt 12?